lib/Conversion/TorchToLinalg/CustomOpExample.cpp

```diff
@@ -0,0 +1,55 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#include "torch-mlir/Conversion/TorchToLinalg/TorchToLinalg.h"
+
+#include "../PassDetail.h"
+#include "PopulatePatterns.h"
+#include "Utils.h"
+#include "mlir/IR/Matchers.h"
+#include "torch-mlir/Conversion/Utils/Utils.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchDialect.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+#include "torch-mlir/Dialect/Torch/Utils/TorchUpstream.h"
+#include "torch-mlir/Dialect/Torch/Utils/Utils.h"
+
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+namespace {
+class ConvertCustomOpExample
+  : public OpConversionPattern<_TorchMlirCustomOpExampleNopOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult matchAndRewrite(_TorchMlirCustomOpExampleNopOp op,
+                                OpAdaptor adaptor,
+                                ConversionPatternRewriter &rewriter
+                                ) const override {
+    // Type checks.
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+    // Since the example op does nothing, we simply replace the uses of the
+    // return value with its argument, then remove the op.
+    op->replaceAllUsesWith(op->getOperands());
+    op->erase();
+
+    return success();
+  }
+};
+} // namespace
+
+void mlir::torch::torch_to_linalg::populateCustomOpExamplePatternsAndLegality(
+    TypeConverter &typeConverter,
+    RewritePatternSet &patterns,
+    ConversionTarget &target) {
+  MLIRContext *context = patterns.getContext();
+  target.addIllegalOp<_TorchMlirCustomOpExampleNopOp>();
```

Comment:
TIL about C++ reserved name formats.

We don't actually control the extension naming scheme. Obviously if someone's extending torch-mlir, they probably have some influence into the upstream extension package, but it still seems a bit wrong to have our code willingly generate illegal names under otherwise legal conditions.

I'll take an alternate view: what's the right thing to do if we *do* encounter a package with a leading underscore? Well, one solution is just to always generate legal code. I.e., when translating the Python names to C++, strip leading underscores. Technically, this collapses `_package_name` and `package_name` to `PackageName`, but... I think that's probably the right answer here, since the alternative is illegal C++ code.

The fix for that is a one-line change to `registry.py`. This has the added benefit of *always* preventing illegal names, which seems positive.

(I'll fix the string delimiters in `registry.py` that I overlooked in a bit. Currently debugging a crash that I can't reproduce locally.)

Reply:
What is it on the Torch side that requires the leading _ ? Is it the "namespace" of the extension? The .so file name?

To the degree that the leading _ is mandatory, then removing it should not result in collisions.

Reply:
> What is it on the Torch side that requires the leading _ ? Is it the "namespace" of the extension? The .so file name?

The extension package name, which ends up being concatenated with the op to give a unique op name. The .so file can be named whatever you wish, since it's entirely controlled by the extension and its build files.

Realistically, nothing about our name translation system is "mandatory", since it's all one-way. Torch-mlir choose how to change Python names into C++ names, and then those C++ names are used internally in torch-mlir. There's nothing that would prevent us from turning `aten.tanh` into `XXMLIRXXatenXXDOTXXtanhXXMLIRXX`, besides driving everyone batty reading it.

So if we think about the whole name translation system as a map, it might look like this:
```
Python Names -> C++ Names
-------------------------
package.op -> PackageOp
package_foo.op_foo -> PackageFooOpFoo
foo.bar.op -> FooBarOp
```

The case we're talking about is this:
```
_package.op -> PackageOp (proposed, currently illegal as _PackageOp)
package.op -> PackageOp
foo._bar.op -> Foo_BarOp
```

Both `_package.op` and `package.op` are technically legal Python names which can register op extensions, so under the current system, they can generate illegal C++ identifiers (without warning). I suppose we could alter the renaming scheme to change leading underscores into something else like `_package.op -> UnderscorePackageOp`, but this seems to violate the principle of least surprise.

At the end of the day, the problem is that this part of the legal Python namespace is a superset of the legal C++ names, so there's always going to be a bit of a rub. We kind of have to choose one of the three:

1. Allow for the possibility of illegal C++ names.
2. Allow for the possibility of namespace collisions.
3. Mangle the python names

My opinion was that 2. is the least bad/intrusive, but I'm happy to reconsider.

Reply:
These two quotes seem to be saying opposite things:

1.
> > What is it on the Torch side that requires the leading _ ? Is it the "namespace" of the extension? The .so file name?
> The extension package name, which ends up being concatenated with the op to give a unique op name. 

2.
> Both _package.op and package.op are technically legal Python names which can register op extensions


Is the leading `_` mandatory?

- If it is mandatory, then dropping the leading _ should never result in a collision
- If it is not mandatory, then why don't we just name our package without it?

Either way, I think that your option 2 makes sense.

Reply:
Sorry, I'll try to be a more clear:

PyTorch extensions aren't necessarily written for torch-mlir, so I'm assuming we don't control the package naming scheme in the general case. The leading `_` is not mandatory for *our* example package and we could definitely remove it, but there's nothing that prevents anyone from developing a PyTorch extension that is named that way. If they do, we don't have much choice in what is fed to us. The name will be whatever the name is. So in that sense, we only have a choice about what to do with what's coming down the pipe.

Honestly, I don't have any kind of census about how often leading-underscore names appear in PyTorch op extensions (I suspect it is not common), so this could really all be unimportant either way.

Reply:
Sounds good. Can always come back with another fix if it becomes a problem.